### PR TITLE
Make it possible to broadcast events to (web) clients

### PR DESCRIPTION
**Pull request by ChristophWurst**
Best reviewed together with https://github.com/nextcloud/push/pull/27

---

Interesting reads on the topic

https://symfony.com/blog/symfony-gets-real-time-push-capabilities
https://github.com/dunglas/mercure
https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events
https://laravel.com/docs/5.7/broadcasting#introduction

### CODE REVIEW

File: core/Command/Broadcast/Test.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
+<?php declare(strict_types=1);
+
```
Comment (rullzer) [inline core/Command/Broadcast/Test.php]: where is your AGPL header ;)

**Review by rullzer — APPROVED** [review]
Let do this!

**Comment by skjnldsv** [comment]
Any news? :)

**Comment by ChristophWurst** [comment]
> Any news? :)

Time to review :)

File: lib/private/Broadcast/Events/BroadcastEvent.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
+<?php declare(strict_types=1);
+
```
Comment (rullzer) [inline lib/private/Broadcast/Events/BroadcastEvent.php]: also here :wink:

File: lib/public/Broadcast/Events/IBroadcastEvent.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?php declare(strict_types=1);
+
```
Comment (rullzer) [inline lib/public/Broadcast/Events/IBroadcastEvent.php]: and here

File: lib/public/EventDispatcher/ABroadcastedEvent.php [hunk 4]
```diff
@@ -0,0 +1,46 @@
+<?php declare(strict_types=1);
+
```
Comment (rullzer) [inline lib/public/EventDispatcher/ABroadcastedEvent.php]: and here

**Comment by nickvergessen** [comment]
Didn't test, but do you know if this works with http2?
The long polling we currently use in Talk breaks the Nextcloud on http2, because the connection is always kept open.

**Comment by ChristophWurst** [comment]
I only tested on nginx with http2 and could not see any problems with that. And the fallback does not use long polling but just returns immediately. I think long polling is also problematic in Nextcloud if you use sqlite as one request can block all others do to the single user restriction.

**Comment by ChristophWurst** [comment]
> Didn't test, but do you know if this works with http2?
> The long polling we currently use in Talk breaks the Nextcloud on http2, because the connection is always kept open.

Turned out this was caused by a misconfigured apache.

Review time :)

**Review by rullzer — APPROVED** [review]

**Comment by ChristophWurst** [comment]
@daita @juliushaertl @nickvergessen @skjnldsv mind giving this a review? Then I can continue with the part in Push :)

The sooner we get this in the more time there is to experiment and iterate on the API if necessary.

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**